File: mergeSort.py
import random
n=10000

def sort(arr,l,m,tamSubListas,tamSubListas2):
    arr3=[]
    arr3=arr[:]
    ln=l
    m0=m
    l0=l

    for i in range(0,(m-l)*2):
        if arr3[l]<=arr3[m]:
            arr[ln]=arr3[l]
            ln=ln+1
            if l+1<m0:
                l=l+1
            else:
                for j in range(m,m0+tamSubListas2):#faltan 2do lado de la lista#*******
                    arr[ln]=arr3[m]
                    m=m+1
                    ln=ln+1   
                return arr 
        else:
            arr[ln]=arr3[m]
            ln=ln+1
            if m+1< m0+tamSubListas2:
                m=m+1
            else:
                for j in range(l,l0+tamSubListas):
                    arr[ln]=arr3[l]
                    l=l+1
                    ln=ln+1
                return arr
    return arr

def mergeSort(arr):
    i=0
    j=0
    k=0
    it=0
    entero=0
    residuo=0
    arr2=[]
    #arr2.append(arr[0])
    arr2=arr[:]
#####
    #if n==1:break
    while 2**i<n:
        i=i+1#numero de veces que ejecuta el ciclo principal
    for j in range(0,i):
        k=2**j#tamano de sub listas
        entero=n//k #numero de sub listas
        residuo=n%k# tamano de lista residuo
        l=0
        for j in range(0, entero/2):
            print(l) 
            arr2=sort(arr2,l,l+k,k,k) 
            l=l+2*k 
           # print("arreglo", arr2)###hasta aqui hace para potencias de 2
    print("residuo", residuo)
    if residuo>0:
        arr2=sort(arr2,0,k,k,residuo)
    print(arr2) 
#####
arr=[]
for i in range(n):
    v=random.randint(0,n)
    arr.append(v)
print(arr)    
mergeSort(arr)

File: mergeSort.c
#include<stdio.h>
#include<string.h>
#include<stdlib.h>
#include<time.h>
#include<math.h>
#define N 10000

int potenciaEnteros(int base, int exponente)
{
	if(exponente==0)return 1;
	int resultado=1,i=0;
	for(i=0;i<exponente;i++)resultado*=base;
	return resultado;
}

int sort(int arr[N],int l,int m, int tamSubListas, int tamSubListas2,int operaciones[])
{	
	int m0=0,l0=0,ln=0,i=0,j=0,c=0;
	ln=l;
    m0=m;
	l0=l;
	int arr3[N];

	for(c=0;c<N;c++)arr3[c]=arr[c];
	

	for(i=0;i<(m-l)*2;i++)
	{
		if(arr3[l]<=arr3[m])
		{
			arr[ln]=arr3[l];
			operaciones[0]++;
			ln++;
			if(l+1<m0)
			{
				l++;
			}
			else
			{
				for(j=m;j<m0+tamSubListas2;j++)
				{
					operaciones[0]++;
					arr[ln]=arr3[m];
					ln++;
					m++;
				}
			return 0;
			}
		}
		else
		{//#
			arr[ln]=arr3[m];
			ln++;
			operaciones[0]++;
			if(m+1<m0+tamSubListas2)
			{
				m++;
			}
			else
			{
				for(j=l;j<l0+tamSubListas;j++)
				{
					operaciones[0]++;
					arr[ln]=arr3[l];
					l++;
					ln++;
				}
				return 0;
			}
		}//#
	}
	return 0;
}

void mergeSort(int arr[N],int iteraciones[], int operaciones[])
{	
	int i=0,j=0,j1=0,k=0,c=0,l=0,z=0;
  	int entero=0,residuo=0;
	int arr2[N];
	for (c=0;c<N;c++)
	{
		arr2[c]=arr[c];
		//printf("%d",arr2[c]);
	}
    
	
	while(potenciaEnteros(2,i)<N)
	{
		i++;//numero de veces que ejecuta el ciclo principal
	}
	
	for(j=0;j<i;j++)
	{
		k=potenciaEnteros(2,j);
		entero=N/k;
		residuo=N%k;
		l=0;
		for(j1=0;j1<entero/2;j1++)
		{
        	//printf("%d",l);
			iteraciones[0]++;
			sort(arr,l,l+k,k,k,operaciones);
			l=l+2*k;
        	//for(z=0;z<N;z++)printf("%d",arr[z]);//potencias de dos
            //printf("\n");
		}
	}
	
	if(residuo>0)
	{	
		iteraciones[0]++;
	 	sort(arr,0,k,k,residuo,operaciones);
	}

//printf("Ordenado");
//for(i=0;i<N;i++)printf("%d ",arr[i]);      
}

int main()
{   
	int arr[N]={0};
	int i=0,v=0;
	int iteraciones[1];
    int operaciones[1];
    iteraciones[0]=0;
	operaciones[0]=0;
	srand(time(NULL));
	for(i=0;i<N;i++)
	{
		v=rand()%N;
		arr[i]=v;
		//printf("%d ",v);
	}
    printf("\n");
    mergeSort(arr,iteraciones,operaciones);	
	printf("iteraciones %d",iteraciones[0]);
	printf("operaciones %d",operaciones[0]);
}
